Conversion error in case of nested field of array<Struct> records

## Actual Behavior (Mandatory)

When the avro file contains a nested field of array<Struct> records.
Thanks to @rmoff for the feedback.

```bash
 [2019-03-11 10:52:26,053] ERROR WorkerSinkTask{id=Neo4jSinkConnector-twitter-0} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted. (org.apache.kafka.connect.runtime.WorkerSinkTask)
 org.apache.kafka.connect.errors.DataException: Exception thrown while processing field 'HashtagEntities'
    at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertStruct(AbstractConverter.java:259)
    at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convert(AbstractConverter.java:78)
    at streams.kafka.connect.sink.Neo4jService.write(Neo4jService.kt:69)
    at streams.kafka.connect.sink.Neo4jService.access$write(Neo4jService.kt:19)
    at streams.kafka.connect.sink.Neo4jService$writeData$2$invokeSuspend$$inlined$flatMap$lambda$1.invokeSuspend(Neo4jService.kt:93)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:32)
    at kotlinx.coroutines.DispatchedTask$DefaultImpls.run(Dispatched.kt:235)
    at kotlinx.coroutines.DispatchedContinuation.run(Dispatched.kt:81)
    at kotlinx.coroutines.scheduling.Task.run(Tasks.kt:94)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:586)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:732)
 Caused by: org.neo4j.driver.v1.exceptions.ClientException: Unable to convert org.apache.kafka.connect.data.Struct to Neo4j Value.
    at org.neo4j.driver.v1.Values.value(Values.java:130)
    at org.neo4j.driver.v1.Values.value(Values.java:232)
    at org.neo4j.driver.v1.Values.value(Values.java:115)
    at streams.kafka.connect.sink.ValueConverter.setValue(ValueConverter.kt:20)
    at streams.kafka.connect.sink.ValueConverter.setArray(ValueConverter.kt:45)
    at streams.kafka.connect.sink.ValueConverter.setArray(ValueConverter.kt:14)
    at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertStruct(AbstractConverter.java:248)
    ... 11 more
```

## How to Reproduce the Problem
Use the `Twitter as a source of data`

```bash
echo '{
    "name": "twitter_source_avro",
    "config": {
      "tasks.max" : "1",
      "key.converter": "io.confluent.connect.avro.AvroConverter",
      "key.converter.schema.registry.url": "http://schema_registry:8081",
      "value.converter": "io.confluent.connect.avro.AvroConverter",
      "value.converter.schema.registry.url": "http://schema_registry:8081",
      "kafka.delete.topic": "twitter_deletes_avro",
      "twitter.oauth.consumerKey": "get yours from dev.twitter.com",
      "twitter.oauth.consumerSecret": "get yours from dev.twitter.com",
      "twitter.oauth.accessToken": "get yours from dev.twitter.com",
      "twitter.oauth.accessTokenSecret": "get yours from dev.twitter.com",
      "kafka.status.topic": "twitter_avro",
      "connector.class": "com.github.jcustenborder.kafka.connect.twitter.TwitterSourceConnector",
      "process.deletes": true, 
      "filter.keywords": "java"
    }
  }
  ' | http POST localhost:8083/connectors
```

```bash
echo '{
  "name": "Neo4jSinkConnector-twitter",
  "config": {
    "topics": "twitter_avro",
    "connector.class": "streams.kafka.connect.sink.Neo4jSinkConnector",
    "errors.retry.timeout": "-1",
    "errors.retry.delay.max.ms": "1000",
    "errors.tolerance": "all",
    "errors.log.enable": true,
    "errors.log.include.messages": true,
    "neo4j.server.uri": "bolt://neo4j:7687",
    "neo4j.authentication.basic.username": "neo4j",
    "neo4j.authentication.basic.password": "connect",
    "neo4j.topic.cypher.twitter_avro": "MERGE (u:User{ID: event.User.Id}) ON CREATE SET u.Name = event.User.Name, u.ScreenName = event.User.ScreenName, u.Location = event.Location MERGE (t:Tweet{ID: event.Id}) ON CREATE SET t.text = event.Text MERGE (u)-[:TWEETED]->(t)"
  }
} ' | http POST localhost:8083/connectors
```

--- a/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/Neo4jService.kt
+++ b/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/Neo4jService.kt
@@ -18,6 +18,7 @@ import org.slf4j.LoggerFactory
 import streams.utils.StreamsUtils
 import streams.utils.retryForException
 import org.apache.kafka.connect.errors.ConnectException
+import org.neo4j.driver.v1.exceptions.Neo4jException
 import java.util.concurrent.TimeUnit
 
 
@@ -28,6 +29,10 @@ class Neo4jService(private val config: Neo4jSinkConnectorConfig) {
     private val log: Logger = LoggerFactory.getLogger(Neo4jService::class.java)
 
     private val driver: Driver
+
+    private val skipNeo4jErrors = listOf("Neo.ClientError.Statement.PropertyNotFound",
+            "Neo.ClientError.Statement.SemanticError",
+            "Neo.ClientError.Statement.ParameterMissing")
 
     init {
         val configBuilder = Config.build()
@@ -77,17 +82,29 @@ class Neo4jService(private val config: Neo4jSinkConnectorConfig) {
                 runBlocking {
                     retryForException<Unit>(exceptions = arrayOf(ClientException::class.java, TransientException::class.java),
                             retries = config.retryMaxAttempts, delayTime = 0) { // we use the delayTime = 0, because we delegate the retryBackoff to the Neo4j Java Driver
-                        session.writeTransaction {
-                            val result = it.run(query, data)
-                            if (log.isDebugEnabled) {
-                                log.debug("Successfully executed query: `$query`. Summary: ${result.summary()}")
+                        try {
+                            session.writeTransaction {
+                                val result = it.run(query, data)
+                                if (log.isDebugEnabled) {
+                                    log.debug("Successfully executed query, summary: ${result.summary()}")
+                                }
+                            }
+                        } catch (neoException: Neo4jException) {
+                            if (skipNeo4jErrors.contains(neoException.code())) {
+                                log.info("Skip query: `$query`. Error message `${neoException.message}`. Error code: `${neoException.code()}`")
+                            } else {
+                                throw neoException
                             }
                         }
                     }
                 }
             } catch (e: Exception) {
                 if (log.isDebugEnabled) {
-                    log.debug("Exception `${e.message}` while executing query: `$query`, with data: `$data`")
+                    if (e is Neo4jException) {
+                        log.debug("Exception `${e.message}` with code `${e.code()}` while executing query `$query`, with data `$data`")
+                    } else {
+                        log.debug("Exception `${e.message}` while executing query `$query`, with data `$data`")
+                    }
                 }
                 throw e
             }

--- a/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/ValueConverter.kt
+++ b/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/ValueConverter.kt
@@ -1,107 +1,130 @@
 package streams.kafka.connect.sink
 
 import com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter
+import com.google.common.base.Preconditions
 import org.apache.kafka.connect.data.Schema
 import org.apache.kafka.connect.data.Struct
+import org.apache.kafka.connect.errors.DataException
 import org.neo4j.driver.v1.Values
 import java.math.BigDecimal
+import java.math.BigInteger
 import java.time.LocalTime
 import java.time.ZoneId
 import java.util.*
 import java.util.concurrent.TimeUnit
 
 
-class ValueConverter: AbstractConverter<MutableMap<String, Any>>() {
+class ValueConverter: AbstractConverter<MutableMap<String, Any?>>() {
 
     private val UTC = ZoneId.of("UTC")
 
-    private fun setValue(result: MutableMap<String, Any>?, fieldName: String?, value: Any?) {
-        if (result != null && fieldName != null && value != null) {
-            result[fieldName] = Values.value(value)
+    private fun setValue(result: MutableMap<String, Any?>?, fieldName: String?, value: Any?) {
+        if (result != null && fieldName != null) {
+            result[fieldName] = Values.value(value) ?: Values.NULL
         }
     }
 
-    override fun newValue(): MutableMap<String, Any> {
+    override fun newValue(): MutableMap<String, Any?> {
         return mutableMapOf()
     }
 
-    override fun setBytesField(result: MutableMap<String, Any>?, fieldName: String?, value: ByteArray?) {
+    override fun setBytesField(result: MutableMap<String, Any?>?, fieldName: String?, value: ByteArray?) {
         setValue(result, fieldName, value)
     }
 
-    override fun setStringField(result: MutableMap<String, Any>?, fieldName: String?, value: String?) {
+    override fun setStringField(result: MutableMap<String, Any?>?, fieldName: String?, value: String?) {
         setValue(result, fieldName, value)
     }
 
-    override fun setFloat32Field(result: MutableMap<String, Any>?, fieldName: String?, value: Float?) {
+    override fun setFloat32Field(result: MutableMap<String, Any?>?, fieldName: String?, value: Float?) {
         setValue(result, fieldName, value)
     }
 
-    override fun setInt32Field(result: MutableMap<String, Any>?, fieldName: String?, value: Int?) {
+    override fun setInt32Field(result: MutableMap<String, Any?>?, fieldName: String?, value: Int?) {
         setValue(result, fieldName, value)
     }
 
-    override fun setArray(result: MutableMap<String, Any>?, fieldName: String?, schema: Schema?, array: MutableList<Any?>?) {
-        setValue(result, fieldName, array)
+    override fun setArray(result: MutableMap<String, Any?>?, fieldName: String?, schema: Schema?, array: MutableList<Any?>?) {
+        val convertedArray = array?.map { convertInner(it) }
+        setValue(result, fieldName, convertedArray)
     }
 
-    override fun setTimestampField(result: MutableMap<String, Any>?, fieldName: String?, value: Date?) {
+    override fun setTimestampField(result: MutableMap<String, Any?>?, fieldName: String?, value: Date?) {
         if (value != null) {
             val localDate = value.toInstant().atZone(UTC).toLocalDateTime()
             setValue(result, fieldName, localDate)
+        } else {
+            setNullField(result, fieldName)
         }
 
     }
 
-    override fun setTimeField(result: MutableMap<String, Any>?, fieldName: String?, value: Date?) {
+    override fun setTimeField(result: MutableMap<String, Any?>?, fieldName: String?, value: Date?) {
         if (value != null) {
             val time = LocalTime.ofNanoOfDay(TimeUnit.MILLISECONDS.toNanos(value.time))
             setValue(result, fieldName, time)
+        } else {
+            setNullField(result, fieldName)
         }
     }
 
-    override fun setInt8Field(result: MutableMap<String, Any>?, fieldName: String?, value: Byte?) {
+    override fun setInt8Field(result: MutableMap<String, Any?>?, fieldName: String?, value: Byte?) {
         setValue(result, fieldName, value)
     }
 
-    override fun setStructField(result: MutableMap<String, Any>?, fieldName: String?, value: Struct?) {
+    override fun setStructField(result: MutableMap<String, Any?>?, fieldName: String?, value: Struct?) {
         if (value != null) {
-            val converted = convert(value) as Map<String, Any>
-            setValue(result, fieldName, converted)
+            val converted = convert(value) as MutableMap<Any?, Any?>
+            setMap(result, fieldName, null, converted)
+        } else {
+            setNullField(result, fieldName)
         }
     }
 
-    override fun setMap(result: MutableMap<String, Any>?, fieldName: String?, schema: Schema?, map: MutableMap<Any?, Any?>?) {
-        setValue(result, fieldName, map)
+    override fun setMap(result: MutableMap<String, Any?>?, fieldName: String?, schema: Schema?, map: MutableMap<Any?, Any?>?) {
+        val newMap = map
+                ?.mapKeys { it.key.toString() }
+                ?.mapValues { convertInner(it.value) }
+        setValue(result, fieldName, newMap)
     }
 
-    override fun setNullField(result: MutableMap<String, Any>?, fieldName: String?) {}
+    override fun setNullField(result: MutableMap<String, Any?>?, fieldName: String?) {
+        setValue(result, fieldName, null)
+    }
 
-    override fun setFloat64Field(result: MutableMap<String, Any>?, fieldName: String?, value: Double?) {
+    override fun setFloat64Field(result: MutableMap<String, Any?>?, fieldName: String?, value: Double?) {
         setValue(result, fieldName, value)
     }
 
-    override fun setInt16Field(result: MutableMap<String, Any>?, fieldName: String?, value: Short?) {
+    override fun setInt16Field(result: MutableMap<String, Any?>?, fieldName: String?, value: Short?) {
         setValue(result, fieldName, value)
     }
 
-    override fun setInt64Field(result: MutableMap<String, Any>?, fieldName: String?, value: Long?) {
+    override fun setInt64Field(result: MutableMap<String, Any?>?, fieldName: String?, value: Long?) {
         setValue(result, fieldName, value)
     }
 
-    override fun setBooleanField(result: MutableMap<String, Any>?, fieldName: String?, value: Boolean?) {
+    override fun setBooleanField(result: MutableMap<String, Any?>?, fieldName: String?, value: Boolean?) {
         setValue(result, fieldName, value)
     }
 
-    override fun setDecimalField(result: MutableMap<String, Any>?, fieldName: String?, value: BigDecimal?) {
+    override fun setDecimalField(result: MutableMap<String, Any?>?, fieldName: String?, value: BigDecimal?) {
         setValue(result, fieldName, value)
     }
 
-    override fun setDateField(result: MutableMap<String, Any>?, fieldName: String?, value: Date?) {
+    override fun setDateField(result: MutableMap<String, Any?>?, fieldName: String?, value: Date?) {
         if (value != null) {
             val localDate = value.toInstant().atZone(UTC).toLocalDate()
             setValue(result, fieldName, localDate)
+        } else {
+            setNullField(result, fieldName)
         }
     }
 
+    private fun convertInner(value: Any?): Any? {
+        return when (value) {
+            is Struct, is Map<*, *> -> convert(value)
+            else -> value
+        }
+    }
 }

--- a/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jSinkTaskTest.kt
+++ b/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jSinkTaskTest.kt
@@ -13,6 +13,7 @@ import org.junit.Before
 import org.junit.Test
 import org.neo4j.graphdb.Label
 import org.neo4j.graphdb.Node
+import org.neo4j.graphdb.RelationshipType
 import org.neo4j.harness.ServerControls
 import org.neo4j.harness.TestServerBuilders
 import java.util.*
@@ -55,6 +56,64 @@ class Neo4jSinkTaskTest: EasyMockSupport() {
             .field("latitude", Schema.FLOAT32_SCHEMA)
             .field("longitude", Schema.FLOAT32_SCHEMA)
             .build()
+
+    private val PERSON_SCHEMA_EXT = SchemaBuilder.struct().name("com.example.Person")
+            .field("firstName", Schema.STRING_SCHEMA)
+            .field("lastName", Schema.STRING_SCHEMA)
+            .field("age", Schema.OPTIONAL_INT32_SCHEMA)
+            .field("bool", Schema.OPTIONAL_BOOLEAN_SCHEMA)
+            .field("short", Schema.OPTIONAL_INT16_SCHEMA)
+            .field("byte", Schema.OPTIONAL_INT8_SCHEMA)
+            .field("long", Schema.OPTIONAL_INT64_SCHEMA)
+            .field("float", Schema.OPTIONAL_FLOAT32_SCHEMA)
+            .field("double", Schema.OPTIONAL_FLOAT64_SCHEMA)
+            .field("modified", Timestamp.SCHEMA)
+            .field("visited", SchemaBuilder.array(PLACE_SCHEMA))
+            .build()
+
+
+
+    @Test
+    fun `test array of struct`() {
+        val firstTopic = "neotopic"
+        val props = mutableMapOf<String, String>()
+        props[Neo4jSinkConnectorConfig.SERVER_URI] = db.boltURI().toString()
+        props["${Neo4jSinkConnectorConfig.TOPIC_CYPHER_PREFIX}$firstTopic"] = """
+            CREATE (b:BODY)
+            WITH event.p AS paragraphList, event.ul AS ulList, b
+            FOREACH (paragraph IN paragraphList | CREATE (b)-[:HAS_P]->(p:P{value: paragraph.value}))
+
+            WITH ulList, b
+            UNWIND ulList AS ulElem
+            CREATE (b)-[:HAS_UL]->(ul:UL)
+
+            WITH ulElem, ul
+            UNWIND ulElem.value AS liElem
+            CREATE (ul)-[:HAS_LI]->(li:LI{value: liElem.value, class: liElem.class})
+        """.trimIndent()
+        props[Neo4jSinkConnectorConfig.AUTHENTICATION_TYPE] = AuthenticationType.NONE.toString()
+        props[Neo4jSinkConnectorConfig.BATCH_SIZE] = 2.toString()
+        props[SinkTask.TOPICS_CONFIG] = "$firstTopic"
+
+        val task = Neo4jSinkTask()
+        task.initialize(mock<SinkTaskContext, SinkTaskContext>(SinkTaskContext::class.java))
+        task.start(props)
+        val input = listOf(SinkRecord(firstTopic, 1, null, null, PERSON_SCHEMA, ValueConverterTest.getTreeStruct(), 42))
+        task.put(input)
+        db.graph().beginTx().use {
+            assertEquals(1, db.graph().findNodes(Label.label("BODY")).stream().count())
+            assertEquals(2, db.graph().findNodes(Label.label("P")).stream().count())
+            assertEquals(2, db.graph().findNodes(Label.label("UL")).stream().count())
+            assertEquals(4, db.graph().findNodes(Label.label("LI")).stream().count())
+
+            assertEquals(2, db.graph().execute("MATCH (b:BODY)-[r:HAS_P]->(p:P) RETURN COUNT(r) AS COUNT").columnAs<Long>("COUNT").next())
+            assertEquals(2, db.graph().execute("MATCH (b:BODY)-[r:HAS_UL]->(ul:UL) RETURN COUNT(r) AS COUNT").columnAs<Long>("COUNT").next())
+            assertEquals(4, db.graph().execute("MATCH (ul:UL)-[r:HAS_LI]->(li:LI) RETURN COUNT(r) AS COUNT").columnAs<Long>("COUNT").next())
+
+            assertEquals(1, db.graph().execute("MATCH (li:LI{class:['ClassA', 'ClassB']}) RETURN COUNT(li) AS COUNT").columnAs<Long>("COUNT").next())
+        }
+    }
+
 
     @Test
     fun `should insert data into Neo4j`() {

--- a/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/ValueConverterTest.kt
+++ b/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/ValueConverterTest.kt
@@ -1,0 +1,121 @@
+package streams.kafka.connect.sink
+
+import org.apache.kafka.connect.data.Schema
+import org.apache.kafka.connect.data.SchemaBuilder
+import org.apache.kafka.connect.data.Struct
+import org.junit.Test
+import org.neo4j.driver.v1.Value
+import org.neo4j.driver.v1.Values
+import kotlin.test.assertEquals
+
+class ValueConverterTest {
+
+    @Test
+    fun `should convert tree struct into map of neo4j values`() {
+        // given
+        // this test generates a simple tree structure like this
+        //           body
+        //          /    \
+        //         p     ul
+        //               |
+        //               li
+        val body = getTreeStruct()
+
+        // when
+        val result = ValueConverter().convert(body) as Map<*, *>
+
+        // then
+        val expected = getExpectedMap()
+        assertEquals(expected, result)
+    }
+
+    @Test
+    fun `should convert tree simple map into map of neo4j values`() {
+        // given
+        // this test generates a simple tree structure like this
+        //           body
+        //          /    \
+        //         p     ul
+        //               |
+        //               li
+        val body = getTreeMap()
+
+        // when
+        val result = ValueConverter().convert(body) as Map<*, *>
+
+        // then
+        val expected = getExpectedMap()
+        assertEquals(expected, result)
+    }
+
+    companion object {
+        private val LI_SCHEMA = SchemaBuilder.struct().name("org.neo4j.example.html.LI")
+                .field("value", Schema.OPTIONAL_STRING_SCHEMA)
+                .field("class", SchemaBuilder.array(Schema.STRING_SCHEMA).optional())
+                .build()
+
+        private val UL_SCHEMA = SchemaBuilder.struct().name("org.neo4j.example.html.UL")
+                .field("value", SchemaBuilder.array(LI_SCHEMA))
+                .build()
+
+        private val P_SCHEMA = SchemaBuilder.struct().name("org.neo4j.example.html.P")
+                .field("value", Schema.OPTIONAL_STRING_SCHEMA)
+                .build()
+
+        private val BODY_SCHEMA = SchemaBuilder.struct().name("org.neo4j.example.html.BODY")
+                .field("ul", SchemaBuilder.array(UL_SCHEMA).optional())
+                .field("p", SchemaBuilder.array(P_SCHEMA).optional())
+                .build()
+
+        fun getTreeStruct(): Struct? {
+            val firstUL = Struct(UL_SCHEMA).put("value", listOf(
+                    Struct(LI_SCHEMA).put("value", "First UL - First Element"),
+                    Struct(LI_SCHEMA).put("value", "First UL - Second Element")
+                            .put("class", listOf("ClassA", "ClassB"))
+            ))
+            val secondUL = Struct(UL_SCHEMA).put("value", listOf(
+                    Struct(LI_SCHEMA).put("value", "Second UL - First Element"),
+                    Struct(LI_SCHEMA).put("value", "Second UL - Second Element")
+            ))
+            val ulList = listOf(firstUL, secondUL)
+            val pList = listOf(
+                    Struct(P_SCHEMA).put("value", "First Paragraph"),
+                    Struct(P_SCHEMA).put("value", "Second Paragraph")
+            )
+            val body = Struct(BODY_SCHEMA)
+                    .put("ul", ulList)
+                    .put("p", pList)
+            return body
+        }
+
+        fun getExpectedMap(): Map<String, Value> {
+            val firstULMap = mapOf("value" to listOf(
+                    mapOf("value" to Values.value("First UL - First Element"), "class" to Values.NULL),
+                    mapOf("value" to Values.value("First UL - Second Element"), "class" to Values.value(listOf("ClassA", "ClassB")))))
+            val secondULMap = mapOf("value" to listOf(
+                    mapOf("value" to Values.value("Second UL - First Element"), "class" to Values.NULL),
+                    mapOf("value" to Values.value("Second UL - Second Element"), "class" to Values.NULL)))
+            val ulListMap = Values.value(listOf(firstULMap, secondULMap))
+            val pListMap = Values.value(listOf(mapOf("value" to Values.value("First Paragraph")),
+                    mapOf("value" to Values.value("Second Paragraph"))))
+            val bodyMap = mapOf("ul" to ulListMap, "p" to pListMap)
+            return bodyMap
+        }
+
+        fun getTreeMap(): Map<String, Any?> {
+            val firstULMap = mapOf("value" to listOf(
+                    mapOf("value" to "First UL - First Element", "class" to null),
+                    mapOf("value" to "First UL - Second Element", "class" to listOf("ClassA", "ClassB"))))
+            val secondULMap = mapOf("value" to listOf(
+                    mapOf("value" to "Second UL - First Element", "class" to null),
+                    mapOf("value" to "Second UL - Second Element", "class" to null)))
+            val ulListMap = listOf(firstULMap, secondULMap)
+            val pListMap = listOf(mapOf("value" to "First Paragraph"),
+                    mapOf("value" to "Second Paragraph"))
+            val bodyMap = mapOf("ul" to ulListMap, "p" to pListMap)
+            return bodyMap
+        }
+    }
+
+}
+

